### ISSUE TITLE
Incorrect canonical URL

### ISSUE THREAD

**Issue by otrelin** [bolt/bolt#7552]
In all my pages (of contenttype pages) there is tag `<link rel="canonical">` which contains the URL of homepage.
`<link rel="canonical" href="http://mydomain.com/">`
Because of this, the pages of the site have lost their positions in the search engines.

When I manually set canonical link (field "Canonical Link" from SEO extension), it is displayed correctly.
But why is it wrong by default? I do not want to set it for each page.
Can I somehow remove this tag from the page at all or disable it?
Is it possible to make this tag absent if the field "Canonical Link" is not filled?

By the way, Sitemap extension also show incorrect link for these pages - for all pages of this contenttype displayed link to homepage.

Interestingly, with the pages of other contenttypes everything is ok.

Details
-------

 - Relevant Bolt Version: [ 3.5.4 ]
 - PHP version: [ 7.2 ]
 - Used web server: [ Apache ]

**Comment by xiaohutai** [comment]
What's the correct canonical in your case then?

There's [a `canonical` setting in `config.yml`](https://github.com/bolt/bolt/blob/3.5/app/config/config.yml.dist#L61-L66).

**Comment by otrelin** [comment]
Correct canonical in every page is it's URL (but not the homepage).
`canonical` setting in my `config.yml` is commented out, as you show in your link.

Now I have:
- page `http://mydomain.com/` -> `<link rel="canonical" href="http://mydomain.com/">`
- page `http://mydomain.com/about` -> `<link rel="canonical" href="http://mydomain.com/">`
- page `http://mydomain.com/contact` -> `<link rel="canonical" href="http://mydomain.com/">`
etc.

I want it to be so:
- page `http://mydomain.com/` -> `<link rel="canonical" href="http://mydomain.com/">`
- page `http://mydomain.com/about` -> `<link rel="canonical" href="http://mydomain.com/about">`
- page `http://mydomain.com/contact` -> `<link rel="canonical" href="http://mydomain.com/contact">`

Or just remove this tag from the page at all.

**Comment by xiaohutai** [comment]
By default it should do exactly as what you described. Then something is wrong in your setup, but I don't know what.

**Comment by otrelin** [comment]
What else can change canonical?
It's only on `pages` contenttype. I use [ContentType overrides](https://docs.bolt.cm/3.5/configuration/routing#contenttype-overrides) for `pages` contenttype - this can be?
Dev-team please help!

**Comment by GwendolenLynch** [comment]
I would say with great certainty that this is related to the SEO extension, and not Bolt core, as this works on literally thousands of sites and is covered under unit test.

**Comment by otrelin** [comment]
Well I also thought so. But when I removed SEO extension nothing changed.
And Sitemap extension has the same problem.
So I didn't know to which issue tracker I need to report this so posted it here.

**Comment by GwendolenLynch** [comment]
Oh, here is fine for now … the conversation is in motion.

Disable **all** extensions and test, then reenable one-by-one would be the best way to start. If we can find a way to reproduce we can either help you, or fix a bug if there is one. :smile:

**Comment by bobdenotter** [comment]
> It's only on pages contenttype. I use ContentType overrides for pages contenttype

Please post the relevant section of your `routing.yml` that does that. :-)

**Comment by xiaohutai** [comment]
I've done the following (based on your description):

- Install Bolt 3.5.4
- Install SEO
- Install Sitemap
- Add `pagebinding` in `routing.yml`

I can't reproduce your issue 🤔 

![worksforme](https://user-images.githubusercontent.com/4630335/42502984-bc486c5c-8437-11e8-8fe4-7877a26a7a00.png)

**Comment by otrelin** [comment]
I made multisite on one installation, so contenttype "pages" are different for german and english site (`seiten` and `pages` respectively). The problem is only in these two contenttypes. In all others (there are six) everyting is ok.
```yml
# static pages - DE
pagebindingde:
    path: /{slug}
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: seite
    contenttype: seiten
    host: %DOMAIN_DE%

# static pages - EN
pagebindingen:
    path: /{slug}
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: page
    contenttype: pages
    host: %DOMAIN_EN%
```

Variables `DOMAIN_DE` and `DOMAIN_EN` are set in `.env`.

**Comment by bobdenotter** [comment]
@otrelin You can only have _one_ route that has `path: /{slug}`. I also don't think it should have the `host:` there. Try removing those, and see what happens.

**Comment by otrelin** [comment]
@bobdenotter when I'll remove `host:` I'll can do this trick only for one domain (which will be the first) and in this case German content type `seiten` will be present both in german and english sites.

**Comment by bobdenotter** [comment]
Yes, I believe that's correct. That's how Bolt works, i'm afraid.

**Comment by otrelin** [comment]
Hmm... I've uninstalled all extensions, removed contenttype override from `routing.yml` but it's still the same, even by URL `http://mydomain.com/page/about`
In pages of other contenttypes its still everything is ok.
Very strange...

**Comment by otrelin** [comment]
Where does this tag adds to the page? Maybe I'll try to debug or to find where it rewrites.

**Comment by bobdenotter** [comment]
@otrelin It does this in `function addSnippets()` in `src/EventListener/SnippetListener.php`. See also `src/Routing/Canonical.php` for how it's set.

**Comment by bobdenotter** [comment]
That said, I'm _quite_ certain it does work as it should under normal circumstances. Are you certain the routing doesn't mess with it? Check the "Routing" panel in the profiler toolbar for details on which route was used. 

Did you clear the cache, after removing the override?

**Comment by otrelin** [comment]
> Check the "Routing" panel in the profiler toolbar for details on which route was used.

@bobdenotter You are right.
My first route rules are:
```yml
homepagede:
    path: /
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: seite
        slug: homepage
    contenttype: seiten
    host: %DOMAIN_DE%

homepageen:
    path: /
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: page
        slug: homepage
    contenttype: pages
    host: %DOMAIN_EN%
```
So for every pages uses this rule (for german and english domain respectively). When I delete this rules, canonical URL shows correct.
The question is - how to set route rules for homepage when GET is empty? And where to insert it?

**Comment by otrelin** [comment]
And it's also interesting - why for pages uses the `homepageen` route, and the content is displayed on the `pagebindingen` route? The content of the pages is shown correctly.

My `routing.yml`:
```yml
# ------------------------------------------------------------------------------
# Place your own routes here, that have a HIGHER priority than the default routes.

homepagede:
    path: /
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: seite
        slug: homepage
    contenttype: seiten
    host: %DOMAIN_DE%

homepageen:
    path: /
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: page
        slug: homepage
    contenttype: pages
    host: %DOMAIN_EN%

# ------------------------------------------------------------------------------
# These are the routes defining the default behaviour of Bolt's frontend.
# Changing these might give unexpected results or even break your website. If
# possible, only add NEW routes above or below these.

homepage:
    path: /
    defaults:
        _controller: controller.frontend:homepage

search:
    path: /search
    defaults:
        _controller: controller.frontend:search

preview:
    path: /preview/{contenttypeslug}
    defaults:
        _controller: controller.frontend:preview
    requirements:
        contenttypeslug: controller.requirement:anyContentType
        
contentlink:
    path: /{contenttypeslug}/{slug}
    defaults:
        _controller: controller.frontend:record
    requirements:
        contenttypeslug: controller.requirement:anyContentType

taxonomylink:
    path: /{taxonomytype}/{slug}
    defaults:
        _controller: controller.frontend:taxonomy
    requirements:
        taxonomytype: controller.requirement:anyTaxonomyType

contentlisting:
    path: /{contenttypeslug}
    defaults:
        _controller: controller.frontend:listing
    requirements:
        contenttypeslug: controller.requirement:pluralContentTypes

# ------------------------------------------------------------------------------
# Place your own routes here, that have a LOWER priority than the default routes.

# static pages - DE
pagebindingde:
    path: /{slug}
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: seite
    contenttype: seiten
    host: %DOMAIN_DE%

# static pages - EN
pagebindingen:
    path: /{slug}
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: page
    contenttype: pages
    host: %DOMAIN_EN%
...
```

**Comment by hansgrinwis** [comment]
Hi @otrelin,

My routing knowledge is not super. These are my thoughts, anyway...

```
homepagede:
    path: /
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: seite
        slug: homepage
    contenttype: seiten
    host: %DOMAIN_DE%
```
`contenttype: seiten` will: 
_Flag this route as "preferred" (canonical) for that ContentType [...]._
(see https://docs.bolt.cm/3.5/configuration/routing)

So this will set the canonical to `/` for all `seiten`.

Removing `contenttype: seiten` from `homepagede:` and `contenttype: pages` from `homepageen:` might solve your last question, too.

**Comment by otrelin** [comment]
Hi @hansgrinwis,

When I removed `contenttype: seiten` from `homepagede` and `contenttype: pages` from `homepageen` all my pages began use routes `pagebindingde` and `pagebindingen` respectively.
The canonical url for the start page is `http://sitename.com/homepage` - according to the `pagebindingde` route (although I want the url `http://sitename.com/`).
The question arises - maybe routes `homepagede` and `homepageen` are not needed? I delete them and get an error when opening the start page `Template "index.twig" is not defined.` - its get it from config but not from routes - not that I need.
Did not help.

Another way - I also removed `contenttype: seiten` from `pagebindingde` and `contenttype: pages` from `pagebindingen`. Now all the pages began use default system route `contentlink`. And the canonical urls became with the segment `pages` and `seiten` respectively - not that I need. My custom routes are not uses but when I delete it - I get errors again.
Did not help.

Another way - I put the routes `pagebindingde` and `pagebindingen` to the top of the routes before homepage routes. Now all the pages also began use them. I delete routes `homepagede` and `homepageen` - again the error when opening the main page. I turn them back and get the first variant - the canonical url of homepage is `http://sitename.com/homepage` - do not that I need.

So this is very strange situation - the routes `homepagede` and `homepageen` are not used (judging by the developer panel), but without them the main page does not work. It turns out that they are still needed, but why then they do not appear in the developer panel? And why then the canonical url of the homepage is `http://sitename.com/homepage` but not `http://sitename.com/`?

**Comment by hansgrinwis** [comment]
Hi @otrelin,

Again, my routing knowledge is not super.

However, what you are trying to do is, using a route for `seiten` (`pagebindingde`), but one `seite` gets a special treatment. It should not set its canonical to `http://sitename.com/{slug}` but to `http://sitename.com/` I think Bolt has a hard time getting that done if you set `contenttype: seiten` for the `pagebindingde`.

What you might do, as a workaround, is using the `controller.frontend:homepage` controller, create an `index.twig` and with a `{% setcontent record=seite/homepage %}` get your record data and with `{% include 'record.twig' %}` render your page.

Good luck!

**Comment by otrelin** [comment]
Hi @hansgrinwis,
Thank you for your help!

But it's also did not work as I need.

When I deleted routes `homepagede` and `homepageen` my start page (in both domains) began use default system route `homepage`. Other pages use my custom routes `pagebindingde` and `pagebindingen` as before.

But with route `homepage` and using `controller.frontend:homepage` record for start page takes from general config "homepage" - from one the same page (for example seite/homepage) both for english and german site. So page meta-title and meta-description show in german (for both languages) and canonical url still `http://sitename.com/{slug}`. When I comment config "homepage" - there is no meta-title and meta-description for the start page.

Content of the page I get as you described:
`{% setcontent record = 'seite/homepage' %}` - for german
`{% setcontent record = 'page/homepage' %}` - for english
and it's show correct. But for get meta-title and meta-description seems variable `record` did not set.

So the problem is still actual.
You exactly described what I want to do:
> However, what you are trying to do is, using a route for `seiten` (`pagebindingde`), but one `seite` gets a special treatment. It should not set its canonical to `http://sitename.com/{slug}` but to `http://sitename.com/`.

**Comment by hansgrinwis** [comment]
Hi @otrelin,

> When I deleted routes homepagede and homepageen my start page (in both domains) began use default system route homepage. Other pages use my custom routes pagebindingde and pagebindingen as before.

That's good!

> But with route homepage and using controller.frontend:homepage record for start page takes from general config "homepage" - from one the same page (for example seite/homepage) both for english and german site. So page meta-title and meta-description show in german (for both languages) and canonical url still http://sitename.com/{slug}. When I comment config "homepage" - there is no meta-title and meta-description for the start page.

Try deleting `homepage: seite/homepage` from your `config.yml`. That way Bolt is not able to understand you are using a `seite` for your homepage and cannot set the canonical to `http://sitename.com/{slug}`. It will not try to retrieve the homepage record. We will do that _inside_  `index.twig`.

Then, in `index.twig`, try to determine which language should be displayed: 
```
{% set temp = canonical() %} 
{% if '.de' in temp %}
    {% setcontent record = 'seite/homepage' %}
{% else %}
    {% setcontent record = 'page/homepage' %}
{% endif %}
```
This code is not tested, so may contain bugs. There might be better ways to detect the domain.

> But for get meta-title and meta-description seems variable record did not set.

I do not understand what you mean here. Sorry.

Good luck again!

**Comment by otrelin** [comment]
@hansgrinwis, yes, I did exactly as you described.

>> But for get meta-title and meta-description seems variable record did not set.
>
> I do not understand what you mean here. Sorry.

I mean, when I set content to variable `record` inside `index.twig` then I have correct data for page content, but I have no meta-title and meta-description for this page, which I entered in backend with SEO Extension.

So, I guess, value in my `record` variable is visible only inside `index.twig` template and in `_master.twig` partial it is empty (in `_master.twig` I call methods `seo.title()` and `seo.metatags()`).
